diff --git a/platforms/chibios/boards/common/ld/STM32H750xB.ld b/platforms/chibios/boards/common/ld/STM32H750xB.ld
index b99c073d44..65724b7e31 100644
--- a/platforms/chibios/boards/common/ld/STM32H750xB.ld
+++ b/platforms/chibios/boards/common/ld/STM32H750xB.ld
@@ -97,48 +97,5 @@ REGION_ALIAS("HEAP_RAM", ram0);
 /* Stack rules inclusion.*/
 INCLUDE rules_stacks.ld
 
-/*===========================================================================*/
-/* Custom sections for STM32H7xx.                                            */
-/* SRAM3 is assumed to be marked non-cacheable using MPU.                    */
-/*===========================================================================*/
-
-/* RAM region to be used for nocache segment.*/
-REGION_ALIAS("NOCACHE_RAM", ram3);
-
-/* RAM region to be used for eth segment.*/
-REGION_ALIAS("ETH_RAM", ram3);
-
-SECTIONS
-{
-    /* Special section for non cache-able areas.*/
-    .nocache (NOLOAD) : ALIGN(4)
-    {
-        __nocache_base__ = .;
-        *(.nocache)
-        *(.nocache.*)
-        *(.bss.__nocache_*)
-        . = ALIGN(4);
-        __nocache_end__ = .;
-    } > NOCACHE_RAM
-
-    /* Special section for Ethernet DMA non cache-able areas.*/
-    .eth (NOLOAD) : ALIGN(4)
-    {
-        __eth_base__ = .;
-        *(.eth)
-        *(.eth.*)
-        *(.bss.__eth_*)
-        . = ALIGN(4);
-        __eth_end__ = .;
-    } > ETH_RAM
-}
-
-/* Code rules inclusion.*/
-INCLUDE rules_code.ld
-
-/* Data rules inclusion.*/
-INCLUDE rules_data.ld
-
-/* Memory rules inclusion.*/
-INCLUDE rules_memory.ld
-
+/* Generic rules inclusion.*/
+INCLUDE rules.ld
