diff --git a/quantum/filesystem/rules.mk b/quantum/filesystem/rules.mk
index 6b15160308..5cc8dda6c3 100644
--- a/quantum/filesystem/rules.mk
+++ b/quantum/filesystem/rules.mk
@@ -1,25 +1,25 @@
 # Filesystem Configurables
-VALID_FILESYSTEM_DRIVERS := lfs_spi_flash
+VALID_FILESYSTEM_DRIVERS := lfs_spi_flash lfs_sdmmc_card
 
 ifeq ($(filter $(FILESYSTEM_DRIVER),$(VALID_FILESYSTEM_DRIVERS)),)
   $(call CATASTROPHIC_ERROR,Invalid FILESYSTEM_DRIVER,FILESYSTEM_DRIVER="$(FILESYSTEM_DRIVER)" is not a valid filesystem driver)
 else
 	OPT_DEFS += -DFILESYSTEM_ENABLE
-	COMMON_VPATH += \
-		$(QUANTUM_DIR)/filesystem
+	COMMON_VPATH += $(QUANTUM_DIR)/filesystem
+	COMMON_VPATH += $(LIB_PATH)/littlefs
+	SRC += \
+		lfs.c \
+		lfs_util.c \
+		fs_lfs_common.c
 
 	ifeq ($(strip $(FILESYSTEM_DRIVER)),lfs_spi_flash)
 		FLASH_DRIVER = spi
 		QUANTUM_LIB_SRC += spi_master.c
-		COMMON_VPATH += \
-			$(LIB_PATH)/littlefs
-		SRC += \
-			lfs.c \
-			lfs_util.c \
-			fs_lfs_common.c \
-			fs_lfs_spi_flash.c
+		SRC += fs_lfs_spi_flash.c
+
+	else ifeq ($(strip $(FILESYSTEM_DRIVER)),lfs_sdmmc_card)
+		SRC += fs_lfs_sdmmc_card.c
+
 	endif
 endif
 
-# TODO: Delete
-OPT_DEFS += -DEXTERNAL_FLASH_SPI_SLAVE_SELECT_PIN=C1
