diff --git a/builddefs/common_features.mk b/builddefs/common_features.mk
index fa9757c6c6..0b66a9030b 100644
--- a/builddefs/common_features.mk
+++ b/builddefs/common_features.mk
@@ -633,7 +633,7 @@ ifeq ($(strip $(XAP_ENABLE)), yes)
         $(error 'XAP_ENABLE = $(XAP_ENABLE)' deprecates 'VIA_ENABLE = $(VIA_ENABLE)'. Please set 'VIA_ENABLE = no')
     endif
 
-    DYNAMIC_KEYMAP_ENABLE := yes
+    DYNAMIC_KEYMAP_ENABLE ?= yes
     FNV_ENABLE := yes
     SECURE_ENABLE := yes
     BOOTMAGIC_ENABLE := yes
